src/rescharstring.rs

```diff
@@ -0,0 +1,151 @@
+//! BER-encoded Restricted Character String types.
+//!
+//! This is an internal module. It’s public types are re-exported by the
+//! parent.
+
+use std::marker::PhantomData;
+use super::{decode, encode};
+use super::OctetString;
+use super::tag::Tag;
+
+
+//------------ LimitedCharSet ------------------------------------------------
+
+/// Trait for the various Restricted Character String types defined in X.680
+pub trait LimitedCharSet {
+
+    /// Implementations must specify their own natural tag
+    const TAG: Tag;
+
+    /// Implementations must verify the contents
+    fn is_allowed<I: Iterator<Item=u8>>(i: &mut I) -> bool;
+}
+
+#[derive(Debug)]
+pub struct CharSetError;
+
+
+//------------ PrintableStringSet --------------------------------------------
+
+/// PrintableString
+///
+/// The printable string is an implementation of the restricted character
+/// string type defined in X.680 that only allows the following characters:
+/// a-z A-Z 0-9 (space) ' ( ) + , _ . / : = ?
+pub type PrintableString = ResCharString<PrintableStringSet>;
+
+#[derive(Debug)]
+pub struct PrintableStringSet;
+
+impl LimitedCharSet for PrintableStringSet {
+
+    const TAG: Tag = Tag::PRINTABLE_STRING;
+
+    fn is_allowed<I: Iterator<Item=u8>>(i: &mut I) -> bool {
+        i.all(|x| {
+                x.is_ascii_alphanumeric() || // A-Z a-z 0-9
+                x == b' ' || x == b'\'' || x == b'(' || x == b')' ||
+                x == b'+' || x == b',' || x == b'-' || x == b'.' ||
+                x == b'/' || x == b':' || x == b'=' || x == b'?'
+        })
+    }
+}
+
+
+//------------ ResCharString -------------------------------------------------
+
+/// A generic Restricted Character String
+///
+/// Contains an OCTET STRING with the actual content.
+///
+/// Specific implementations have to ensure that they verify the content
+/// according to the appropriate limitations, and use the correct Tag.
+///
+#[derive(Clone, Debug)]
+pub struct ResCharString<L: LimitedCharSet> {
```

Comment:
I may have changed my mind and now I tend towards calling the type just `CharString<L>` and the trait just `CharSet`. Alternatively, consider spelling it out `RestrictedCharString`; the length doesn’t matter too much, since people will always use the aliases for concrete character sets.

Reply:
Will go for short names, they could be ambiguous, but I think they make sense within the context.